### PR TITLE
EvalFunc: add automatic differentiation support

### PR DESCRIPTION
**Pull request by akva2**
Sits on top of https://github.com/OPM/IFEM/pull/643

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=245 ifem-navierstokes=197 ifem-boussinesq=99 ifem-burgers=4 ifem-elasticity=150 please

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Review by kmokstad — APPROVED** [review]
Lots of template magic hwew which is hard to follow, but I trust to does the job.
I assume the `if constexpr (std::is_same_v<Scalar,Real>)` statement is used the check for equality of template arguments?

**Comment by akva2** [comment]
yeah. it's a compile time if. std::is_same_v<a,b> checks if a and b are the same types and in that case returns true, else false. since it's compile time, this means only one of the code paths are chosen when instancing the template for that type.

**Comment by kmokstad** [comment]
right, which only could be achived using some #ifdef expressions before modern times. Neat
